File: Scala/src/main/scala/AlgorithmicHeights/BIP.scala
package AlgorithmicHeights

object BIP {
  import scala.annotation.tailrec
  import scala.collection.mutable.{Queue => MutableQueue}

  sealed trait NodeColor
  final case object Red extends NodeColor
  final case object Blue extends NodeColor

  final case class Edge(a: Int, b: Int)

  class Graph(val nrNodes: Int, edgeList: List[Edge]) {
    import Graph.buildAdjacencyList

    private val adjacencyList: Map[Int, List[Int]] = buildAdjacencyList(edgeList)

    def neighbours(node: Int): List[Int] = adjacencyList.getOrElse(node, Nil)
  }

  object Graph {
    private def buildAdjacencyList(edgeList: List[Edge]): Map[Int, List[Int]] =
      edgeList
        .flatMap{ case edge @ Edge(a, b) => List(edge, Edge(b, a)) }
        .groupMap{ case Edge(a, _) => a }{ case Edge(_, b) => b }
  }

  private def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  private def readParameters(line: String): (Int, Int) = convertToIntList(line) match {
    case List(n, e) => (n, e)
    case _ => throw new Exception("Malformed input.")
  }

  private def readEdges(reader: Iterator[String], nrEdges: Int): List[Edge] =
    reader
      .take(nrEdges)
      .map(convertToIntList)
      .collect{ case List(a, b) => Edge(a, b) }
      .toList

  private def readGraphs(reader: Iterator[String], n: Int): List[Graph] = {
    (0 until n).map{ _ =>
      val _: String = reader.next()
      val (nrNodes, nrEdges): (Int, Int) = readParameters(reader.next())
      val edgeList: List[Edge] = readEdges(reader, nrEdges)
      new Graph(nrNodes, edgeList)
    }.toList
  }

  private def getOppositeColor(color: NodeColor) = color match {
    case Red => Blue
    case Blue => Red
  }

  private def findConsistentComponentColoring(graph: Graph, startNode: Int): Option[Map[Int, NodeColor]] = {
    val queue: MutableQueue[Int] = MutableQueue(startNode)

    @tailrec
    def loop(nodeColorsInComponent: Map[Int, NodeColor]): Option[Map[Int, NodeColor]] = {
      if (queue.isEmpty) Some(nodeColorsInComponent)
      else {
        val node: Int = queue.dequeue()
        val nodeColor: NodeColor = nodeColorsInComponent(node)
        val oppositeColor: NodeColor = getOppositeColor(nodeColor)
        val neighbours: List[Int] = graph.neighbours(node)
        val (coloredNeighbours, uncoloredNeighbours): (List[Int], List[Int]) =
          neighbours.partition(nodeColorsInComponent.contains)
        if (coloredNeighbours.map(nodeColorsInComponent(_)).contains(nodeColor)) None
        else {
          uncoloredNeighbours.foreach(queue.enqueue)
          loop(nodeColorsInComponent ++ uncoloredNeighbours.map(neighbour => neighbour -> oppositeColor))
        }
      }
    }

    loop(Map(startNode -> Red))
  }

  private def isBipartite(graph: Graph): Boolean = {
    @tailrec
    def loop(nodeColors: Map[Int, NodeColor], startNode: Int): Boolean = {
      if (startNode >= graph.nrNodes) true
      else if (!nodeColors.contains(startNode)) {
        findConsistentComponentColoring(graph, startNode) match {
          case None => false
          case Some(componentColoring) => loop(nodeColors ++ componentColoring, startNode + 1)
        }
      }
      else loop(nodeColors, startNode + 1)
    }

    loop(Map(), startNode = 1)
  }

  def testBipartiteness(graphs: List[Graph]): List[Boolean] = graphs.map(isBipartite)

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val nrExamples: Int = reader.next().toInt
    val graphs: List[Graph] = readGraphs(reader, nrExamples)
    val result: List[Boolean] = testBipartiteness(graphs)
    println(result.map(if (_) "1" else "-1"))
  }
}

File: OCaml/src/algorithmic_heights/bip.ml
module IntMap = Map.Make(Int)

type node_color = Red | Blue

type edge = { a : int; b : int }

module Graph : sig
    type t
    val create : int -> edge list -> t
    val get_neighbors : t -> int -> int list
    val nr_nodes : t -> int
end = struct
    type t = { nr_nodes : int; adjacency_list : int list IntMap.t }

    let nr_nodes { nr_nodes; _ } = nr_nodes

    let get_neighbors { adjacency_list; _ } node =
        match IntMap.find_opt node adjacency_list with
            | None -> []
            | Some neighbors -> neighbors

    let create nr_nodes edges =
        let add_edge adjacency_list { a; b } =
            let update_function = function
                | None -> Some [b]
                | Some ns -> Some (b :: ns) in
            IntMap.update a update_function adjacency_list in
        let add_edge_both_way adjacency_list { a; b } =
            add_edge (add_edge adjacency_list { a; b }) { a = b; b = a } in
        let adjacency_list = List.fold_left add_edge_both_way IntMap.empty edges in
        { nr_nodes; adjacency_list }

end


let get_opposite_color = function
    | Red -> Blue
    | Blue -> Red


let read_graphs (n : int) : Graph.t list =
    let read_edges (nr_edges : int) : edge list =
        let parse_line line =
            Scanf.sscanf line "%d %d" (fun a b -> { a; b }) in
        let range = Seq.init nr_edges (fun k -> k) in
        List.of_seq @@ Seq.map (fun _ -> parse_line (read_line ())) range in
    let read_graph () =
        ignore (read_line ());
        let (nr_nodes, nr_edges) = Scanf.sscanf (read_line ()) "%d %d" (fun n e -> (n, e)) in
        let edges = read_edges nr_edges in
        Graph.create nr_nodes edges in
    List.(map (fun _ -> read_graph ()) @@ init n (fun _ -> 0))


let find_consistent_component_coloring (graph : Graph.t) (start_node : int) : (node_color IntMap.t) option =
    let queue = Queue.create () in
    Queue.add start_node queue;
    let rec loop node_colors_in_component = match Queue.take_opt queue with
        | None -> Some node_colors_in_component
        | Some node ->
            let color = IntMap.find node node_colors_in_component in
            let opposite_color = get_opposite_color color in
            let neighbors = Graph.get_neighbors graph node in
            let (colored_neighbors, uncolored_neighbors) =
                List.partition (fun n -> IntMap.mem n node_colors_in_component) neighbors in
            if List.exists (fun n -> color = IntMap.find n node_colors_in_component) colored_neighbors then None
            else
                (Queue.add_seq queue (List.to_seq uncolored_neighbors);
                let new_nodes = List.(uncolored_neighbors |> map (fun n -> (n, opposite_color)) |> to_seq) in
                loop (IntMap.add_seq new_nodes node_colors_in_component)) in
    loop (IntMap.singleton start_node Red)


let is_bipartite (graph : Graph.t) : bool =
    let rec loop node_colors start_node =
        if start_node >= (Graph.nr_nodes graph) then true
        else if IntMap.mem start_node node_colors then loop node_colors (start_node + 1)
        else
            match find_consistent_component_coloring graph start_node with
                | None -> false
                | Some component_coloring ->
                    let bindings = IntMap.bindings component_coloring in
                    loop (IntMap.add_seq (List.to_seq bindings) node_colors) (start_node + 1) in
    loop IntMap.empty 1


let () =
    let nr_examples = read_int () in
    let graphs = read_graphs nr_examples in
    let results = List.map is_bipartite graphs in
    results |> List.map (fun res -> if res then "1" else "-1") |> String.concat " " |> print_endline
